hack/generator/pkg/astmodel/type.go

```diff
@@ -27,7 +27,7 @@ type Type interface {
 	AsType(codeGenerationContext *CodeGenerationContext) ast.Expr
 
 	// AsDeclarations renders as a Go abstract syntax tree for a declaration
-	AsDeclarations(codeGenerationContext *CodeGenerationContext, name TypeName, description []string) []ast.Decl
+	AsDeclarations(codeGenerationContext *CodeGenerationContext, name TypeName, description []string, validations []KubeBuilderValidation) []ast.Decl
```

Comment:
I did end up adding `DeclarationContext` (mostly to clean up this method signature, not to move validations into `TypeDefinition`). Going to copy my response from chat into here.

Reply:
So originally I was thinking of/did actually sketch out having the validations only be on `TypeDefinition` and `PropertyDefinition` which is where they need to be at codegen time (for the validation comments), but it made several things much more difficult. (The invariant that is observed in the final version is that `ValidationType` only ever appears directly under `TypeDefinition` or `PropertyDefinition`, by codegen time. So they do end up in the right place eventually!)

The first issue is that during the JSONAST phase the validations are only associated with a type (e.g. you can have an array of validated strings), it's only later that we could make a `TypeDefinition` that could hold them. So we have to represent them in the “type tree” at this point anyway. (We then give them names during the "name CRD types" phase which makes sure that all validations (that aren't directly under properties) have names.)
 
I also found that it made manipulations that respect validations much harder because when they only appear in type/property-definitions you have to examine both the the outer property definition, and also deal with indirections through named types; with the validations appearing in the “type tree” it is easier to deal with them (e.g. in `AllOf` the case of combining a validated and non-validated version of the same type is straightforward).
 
I do think that logically the validations belong to the type itself, even though we can’t render that fact  e.g. it makes more sense that if we replace the type of a property then the validations should be replaced along with the type. (i.e. I view `ValidatedType` as creating a new, restricted type…)